### ISSUE TITLE
NppFluentUI Icons for NppDarkMode

### ISSUE THREAD

**Issue by WildByDesign** [notepad-plus-plus/notepad-plus-plus#9742]
I created this separate issue so that icon related details don't clutter up the main discussion PR #9587 that @adzm has created. Essentially, I don't want to create any kind of noise in that discussion because it is more programming related whereas this is more image/icon related.

**Microsoft Fluent UI System Icons:** https://github.com/microsoft/fluentui-system-icons
Easy to navigate site for Fluent UI icons: https://bennymeg.github.io/ngx-fluent-ui/

### My first WIP:
![NppFluentIcons](https://user-images.githubusercontent.com/26308319/113876962-3d03c080-9786-11eb-82d8-943ffc882f84.png)

The first WIP NppFluentUI Icons for NppDarkMode was based on the regular Fluent UI icons.

### My second WIP:
![NppFluentIcons-filled](https://user-images.githubusercontent.com/26308319/113877139-6ae90500-9786-11eb-8a7c-402ba2f80259.png)

The second WIP focused on the filled Fluent UI icon type in the hopes that it would be softer on the eyes. I also focused more on the icons in the tab bar as well. I got rid of the constant save button icon that is always present on each tab. Instead, I made it so that the icon shows up on the tab only when changes are made, making it clear to the user that the latest changes have not yet been saved.

### Challenges and Issues:

Part of the reason why I made the change on the tab bar with the saved icon is because of a problem where the saved icon was being colored constantly regardless of the color that I was using. I assume this is related to the following issue mentioned over at (https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9587#issuecomment-814289896):

> Unfortunately, since the tab color styling has not really worked in a long
> while due to windows' visual styles, pretty much every theme (including the
> default notepad theme) is using incredibly jarring and garish colors for
> tab styles, like bright red and green backgrounds. Ick.

So the saved icon was being color-shaded red or green depending on whether the tab was in focus or not and I believe that this is something that requires coding changes and cannot be solved with icons in tab bar.

There are also a lot of challenges with scaling SVG to tiny sizes (16x16) and converting to old BMP format. So while I could technically batch convert all icons to make the process easier for myself, the batch converted icons suffered from some imperfections and pixel related issues that show up more at 16x16. So I have to do each icon manually and give each pixel some love.

### Release Challenges and Timing:

I am not a programmer and therefore have no ability to add these icons to Npp properly, such as having them automatically used when a user chooses NppDarkMode. I have only simply overwrote the old icons by same file name and compiled Npp that way to have a working example. A user would need to ability to choose the icons from settings or automatically when NppDarkMode is set.

@adzm I am only a few days or a week away from being able to complete these, at least enough for testing scenarios. Possibly we can try to time the release of these icons with the timing of your NppDarkMode merge. It will be great once your NppDarkMode is merged into official Npp branch to make collaboration easier.

**Comment by adzm** [comment]
This is great work. Would you be able to upload the icons you created
somewhere? There might need to be some magic to get them to work properly
with all the dpi stuff.

**Comment by KnIfER** [comment]
Are they *.bmp icons or *.ico icons?

**Comment by WildByDesign** [comment]
> This is great work. Would you be able to upload the icons you created somewhere? There might need to be some magic to get them to work properly with all the dpi stuff.

> Are they *.bmp icons or *.ico icons?

Sorry for the delay in my response to your questions. Initially, I was using the Npp "Standard icons" format since that is the default. Those are the *.bmp files in which the bottom left pixel is used to determine the transparency for the bmp icon. Many of the icons turned out great, but yet a handful of them had me tied up for days because the bottom left pixel for transparency method is not 100% perfect at all times.

So your question about whether or not the icons are *.bmp icons or *.ico led me down another path and started from scratch with the .ico icon format instead. This has given me much better control over the icons, overall sharpness, and it got rid of all of my previous issues regarding bmp bottom left corner pixel transparency that had me stressed over the past few days. Thank you for asking that question because that has led to much better icons in the end. This also gives us the choice between small and large icons which can benefit accessibility for some users. The scaling between 16x16 and 32x32 is working consistently sharp.

I am about 75% done with the Npp Fluent UI Filled icon set. When I complete them, hopefully tomorrow, I will upload them either to my Github repo or somewhere and I will notify you here in this issue. That way we can start laying out the groundwork for the new icons. After that, I am planning to also make the Npp Fluent UI Regular icon set as well.

It is time consuming, but I am saving all changes made to my SVG files prior to exporting to ICO. All of the time consuming changes are saved in the SVG files so that it will be much easier in the future if we need to make changes and so on.

**Comment by WildByDesign** [comment]
It has taken me longer than I expected, but I have completed all toolbar icons for 32x32 pixel ICO (scales down nicely to 16x16 pixel as well) and ended up completing the NppFluentUI Filled and Regular icon sets.

Click on the screenshot for full clarity to ensure it's viewed at 100%.
![NppFluentIcons-ICO](https://user-images.githubusercontent.com/26308319/114805407-f3f8d100-9d70-11eb-8cb2-21c09639aacd.png)

I did not prepare a screenshot yet for the Regular icon set, but the icons are complete. The majority of the icons are sharp at both 32x32 and 16x16 pixels, although there are a few icons that failed to export as sharp. So I will have to redo the SVG process for several of the icons later and can update those icons with sharper versions later.

The Save Recorded Macro button looks pretty goofy still. There was no FluentUI icon available that was appropriate for that button, so I did my best to create that button myself by combining a Play icon with a Save icon. I will have to redo that button as well when I have more time and better ideas for it.

I don't have time to create a Github repo for these icons at the moment, so I am attaching to this comment.

@adzm 

Download:

[NppFluentUI-Icons.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6314879/NppFluentUI-Icons.zip)

At the moment, if you are building Npp, you can drop one of the icon sets (either Filled or Regular) in the `\PowerEditor\src\icons` directory and rebuild Npp. Personally, I would suggest that the Filled icon set is much better at the moment. The Regular icon set still needs more polishing.

Hopefully somebody can program these into Npp properly so that these icons are used automatically when Dark Mode is used, and color icons when Light Mode is used. I don't have any programming experience and therefore that is above my skills.

**Comment by WildByDesign** [comment]
If anyone wants to test an x86 or x64 Debug binary without having to compile, I've got that so that you can play around with it for a bit.

- Npp Dark Mode implemented by @adzm
- Compiled from @namazso branch (v7.9.5)
- Backup your current notepad++.exe binary
- Drop in notepad++.exe binary to replace

Ensure that you switch from the default Standard icon set in Preferences to either Small icons or Big icons because those two options are the ones which utilize the ICO formatted icons.

[NppFluentUI-Debug-x64.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6318847/NppFluentUI-Debug-x64.zip)
[NppFluentUI-Debug-x86.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6318848/NppFluentUI-Debug-x86.zip)

Cheers!

**Comment by adzm** [comment]
Amazing work, thanks. I'm going to check this out asap. We might need some
more icons for some misc things for dark mode anyway.

**Comment by adzm** [comment]
Do you have the source svg and script for generating these? eventually we might want to generate even higher resolutions for high dpi displays, for example.

**Comment by adzm** [comment]
unless we want to differentiate the hover (on / off) states for the .ico files it might be best just to use a bmp since we can have that include an alpha channel with a bit of finagling I believe. anyway, I'm going to experiment with this over the weekend at least and see if I can get it working as an option. There are some straggler icons/bitmaps that we might want to replace as well for consistency.

**Comment by adzm** [comment]
Let me know if you can share a version with just the bmp files with an alpha channel (or the svg source and perhaps a script you were using to generate them and I can try as well?) Thanks for your efforts again

**Comment by WildByDesign** [comment]
Current status as of 04/22/2021: (view full size for best clarity)
![NppFluentIcons-2021-04-22](https://user-images.githubusercontent.com/26308319/115730184-103dd480-a354-11eb-9017-d4c36ee098a2.png)

Since my last update on these icons a few days ago, I have essentially recreated the majority of them. In particular, I put a lot of work into both the Find and Replace icons. Also, the Synchronize Vertical Scrolling and Synchronize Horizontal Scrolling icons were quite terrible before and the new ones represent "scrolling" much better now. I also put some work into the Record Macro button, and completely recreated the Run a Macro Multiple Times icon.

There are a few icons that I need to re-do the process to achieve a sharper outcome. But the main outstanding thing that I still want to recreate is the Save Current Recorded Macro icon. I will recreate that icon later when I have more time.


> Amazing work, thanks. I'm going to check this out asap. We might need some more icons for some misc things for dark mode anyway.

You're welcome, and thank you as well. It was your work with the Npp Dark Mode has that inspired me to put in the time to get these icons working in Npp. It has taken many more hours than I had expected, but it is absolutely worth it in the end.


> unless we want to differentiate the hover (on / off) states for the .ico files it might be best just to use a bmp since we can have that include an alpha channel with a bit of finagling I believe. anyway, I'm going to experiment with this over the weekend at least and see if I can get it working as an option. There are some straggler icons/bitmaps that we might want to replace as well for consistency.

Please let me know which straggler icons I need to do or point out which section of the Npp interface they are located and I will be happy to replace them with some inverted Fluent UI icons as well.

**Comment by WildByDesign** [comment]
@adzm I have the source SVG files included in the zip archive attached to the following comment: https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9742#issuecomment-820019103
The only negative at the moment is that my source SVG files are kept in the same naming scheme as Fluent UI icons are, while the exported ICO files were renamed to match the Npp icon file names exactly. So you can play around with those for now, but in the future I should probably look into renaming the SVG files as well to match the ICO file names.

Regarding ICO/SVG/BMP formats:

Do we know if Notepad++ could potentially support using SVG images directly as toolbar buttons?

That would be ideal as far as achieving the best scaling and sharpness. But since Npp is rather old code base, I would guess that it does not support SVG directly at the current time.

The problem that I ran into with BMP and transparency is that the alpha transparency of the BMP is what was destroying the sharpness and quality of the icons. For example, outside of Npp, the BMP icons would look sharp and fantastic, meanwhile when Npp loads those same BMP icons and alpha transparency, Npp was killing the overall quality and sharpness of the BMP icons. I ended up spending many days trying different methods for the BMP icons but it all ended up being wasted time.

The ICO icons with transparency ended up much better quality and sharpness in comparison to the BMP alternatives.

So the only way forward would be using ICO format for the time being, and potentially in the future if Npp could support loading SVG images directly in the toolbar and other parts of the UI it would be the best case when it comes to scaling and sharpness.

Anyway, I still have to recreate that Save Current Recorded Macro icon and than I will put together another package with icons and source SVGs, and also some more compiled notepad++.exe binaries for testing.

After that, my plan is to organize the naming scheme of the SVG files to match the output of the Npp icons so that I will be able to automate the process with a script for batch conversion using ImageMagick. So I will clean up all of the SVG naming soon and ensure that we have a more reproducible and automated process.

Here is a current screenshot of the standard 16x16 toolbar icon size:
![NppFluentIcons-2021-04-22-small](https://user-images.githubusercontent.com/26308319/115736377-403ba680-a359-11eb-9733-ef4afa868f19.png)

You can see some sharpness issues still with the Save All icon and also the Show Indent Guide icon. I have tried a few times to achieve better sharpness with both of those icons but the conversion of those two have been problematic. Not all of the Fluent UI icons come as 16x16 and therefore some of them had to be scaled within the SVG. Most of them successfully, but those two are a failure thus far.

**Comment by adzm** [comment]
Great work for sure. I'm looking into adding these into notepad++ now, too. I've had good luck in the past by rasterizing vectors at a larger size and then resampling down to smaller sizes with various filters and resampling algorithms (linear will work best for 16x16 in my experience). I'm curious what @donho thinks about this, since it's a menial task to get these into code, might as well make sure it's done right the first time ;) my experimental approach literally just adds 7000 to every icon resource ID and the handful of bmp you've already created and duplicates them. Though this does add almost 400kb to the binary size unfortunately. But that's a technical detail. You are doing great work here in my opinion, and i've been using these by overwriting the existing ones, and using the darkmode patch, and now I don't want to go back.

**Comment by donho** [comment]
@WildByDesign I can even feel your love on every single pixel. Both are perfect, but I love more your 1st WIP - they are pure elegant lines put together by an artist - you have done an amazing job here! 

> Do we know if Notepad++ could potentially support using SVG images directly as toolbar buttons?

Notepad++ is based on Win32 API, and AFAIK, there's no way to use SVG directly as toolbar buttons. But I may be wrong.

@adzm 
> I'm curious what @donho thinks about this, since it's a menial task to get these into code, might as well make sure it's done right the first time

Sorry, I'm not sure that I follow you - what should we get into the code? If you mean @WildByDesign's icon, there's no doubt to me - they will be part of Notepad++ icon sets. I'm even thinking how to use them in light mode. I'm thinking to replace the current "Small icons" & "Big icons", but with which one? the 1st WIP or the 2nd WIP? And we have to create the possibility to use another icon of close button of tab.

> and now I don't want to go back.

Amen.

**Comment by mere-human** [comment]
What if Notepad++ would support icon customization?
For example, it could have several built-in presets but also would allow setting user own icons.
I know this is not an easy thing to do, but I think it is feasible.
I bet other users would be happy.

If not that, then it would be really nice to see @WildByDesign icons in Notepad++. He did an amazing job!

**Comment by mere-human** [comment]
> and AFAIK, there's no way to use SVG directly as toolbar buttons.

I see several options: 
1. Store the SVG icons in the repository. Convert them to ICO, BMP or whatever raster format we want during the build process. Then use the raster icons in the application.
I've seen this done in other projects.
2. Render SVG icons using some 3rd party library. I don't know if it is even possible. Even if it is, then it would require much effort. So lets not think too much about this option🙂

**Comment by adzm** [comment]
I experimented with using a separate resource dll with the same identifiers
for the icons which allowed me to switch them easily enough. That could be
key for future customization.

Honestly I am thinking we should replace the old ugly big/small icons. Does
anyone actually use those anyway? The benefit of using ico over bmp is the
multiple resolutions which will just get more important as high dpi becomes
more common and more dense.

Since these icons are monochromatic it is not too difficult to invert their
colors at runtime so we can have both light and dark versions with one
derived from the other.

**Comment by donho** [comment]
@mere-human 
> What if Notepad++ would support icon customization?

Yes, it could be in the future version. And furthermore, user should be able to customize the toolbar. So both systems combinate will make a quite complex system.

> Store the SVG icons in the repository. Convert them to ICO, BMP or whatever raster format we want during the build process. Then use the raster icons in the application.

It can be considered, indeed.

**Comment by donho** [comment]
@adzm 
> Honestly I am thinking we should replace the old ugly big/small icons. Does anyone actually use those anyway?

With the quite high amount of people who use Notepad++ daily, these outdated (I prefer to use "outdated" instead of "ugly" :) ) icons are used certainly by some users. They are not many for sure. Even though, any application should keep moving forward. So these outdated icons should be changed.

> The benefit of using ico over bmp is the multiple resolutions which will just get more important as high dpi becomes
more common and more dense.

OK, so let's use ico as our standard.

> Since these icons are monochromatic it is not too difficult to invert their
colors at runtime so we can have both light and dark versions with one
derived from the other.

That's what I was thinking. @WildByDesign could you provide me your 2 set of icons with inversed colour, for both light mode & dark mode please? My idea is to include both in Notepad++ binary, with the following UI for switching among the different sets:

![image](https://user-images.githubusercontent.com/90293/115976764-95520500-a571-11eb-9d94-c8ec504525ff.png)

So while the Dark mode is switched in, the colour of 2 sets will be inverted. What do you think  @WildByDesign & @adzm ?

> I experimented with using a separate resource dll with the same identifiers for the icons which allowed me to switch them easily enough. That could be key for future customization.

Such system can be considered indeed, if we have a more sets of icons coming.

**Comment by WildByDesign** [comment]
@donho I like your UI settings options and agree 100%. Inverting icons to use between light mode and dark mode will be the easy part. I also like your idea of having both FluentUI icons types, filled and regular.

Status update on icons: 

It took me the better part of a week, but I have completed the automation of the icon processing of all SVG icons into ICO format, complete with inverting built into the scripting.

The final thing for me to do is finish of polishing of icons that did not look as sharp. I finally figured out what was causing the issue during conversion and I will have to fix some icons directly within the SVGs.

**Comment by donho** [comment]
@WildByDesign 

OK, sounds great. A PR in progress is here: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9808
The UI will look like:
![image](https://user-images.githubusercontent.com/90293/116172205-95d1d380-a70a-11eb-8582-2580043264cd.png)

Once your light mode icons are ready, I'll add into this PR & merge it into master. So people can test the new look & feel of light mode immediately, and @adzm can use them & revert them in dark mode afterward!

**Comment by dinkumoil** [comment]
@donho 

> OK, so let's use ico as our standard.

Is `ICO` intended to be the **exclusive** new icon format or is it to be an **additional** format? If it will be exclusive this will break lots of existing plugins, so I would like to recommend to make it an **additional** format.

**Comment by donho** [comment]
@dinkumoil 

`ICO` IS the current supported format since it's used for small & big icon. It could be the exclusive format in the future, but not in short term - so users can keep using "Standard icons" if they need the plugin icons on toolbar. And indeed, new `ICO` format should be added for the plugins on toolbar. I'll work on it.

**Comment by dinkumoil** [comment]
@donho 

> ICO IS the current supported format

When I wrote my plugins that add icons to Npp's toolbar I was not successful in using images in `ICO` format. I had to use 256 color palletized `BMP` format instead. The color of the pixel in the upper left corner (could be also the lower left corner, I'm not sure) determines the transparent color.

**Comment by adzm** [comment]
Can you link to the plugin source?

**Comment by dinkumoil** [comment]
@adzm 

> Can you link to the plugin source?

Of course, as an example I provide links to my _ReloadFile_ plugin which is so simple that I considered it not worth to be part of the plugin list. But it is written in Delphi...

1. `NPPM_XXX` event dispatcher: https://github.com/dinkumoil/ReloadFile/blob/master/src/Lib/NppPlugin.pas#L300
2. Overridden virtual method that handles `NPPN_TB_MODIFICATION` event: https://github.com/dinkumoil/ReloadFile/blob/master/src/Main.pas#L135
3. Worker method that actually sets the toolbar icon: https://github.com/dinkumoil/ReloadFile/blob/master/src/Lib/NppPlugin.pas#L386

But maybe I misunderstood something because @donho also wrote

> new ICO format should be added for the plugins on toolbar

so it seems like he is aware that `ICO` format currently doesn't work.

**Comment by WildByDesign** [comment]
**Status Update:**

![NppFluentIcons-status](https://user-images.githubusercontent.com/26308319/116483367-f4a36400-a854-11eb-9c96-b57b4f04a0af.png)

**Sharpness at 16x16:**

I have finally achieved a level of sharpness that I am _almost_ satisfied with for the 16x16 size which is important because I believe that would be the default size for the majority of users.

**Inverting and Shading:**

I have a script that does the inverting of icons successfully and another script that colorizes the icons with a shaded color for icons when certain icons, eg. Save, are not active yet. I just have to adjust the script for the Light Mode shaded icons because they don't look quite as good as Dark Mode icons yet.

**Save Current Recorded Macro:**

I finally had time to create a brand new SVG that comes out sharp and basically just combines two FluentUI icon components into one. This one was driving me nuts for a while.

![NppFluentIcons-macrosave](https://user-images.githubusercontent.com/26308319/116484332-dc344900-a856-11eb-869b-8b282db7af8a.png)

**Release ETA:**

The scripting, SVGs, and file/folder name structure is complete. I put a crazy amount of time directly into the SVGs so that it will save us time in the future. A few icons here and there may need some extra polish, however, the structure is itself complete. 

I plan on having icons which have a 32x32 icon layer and a 16x16 icon layer. I have already tested this and Npp utilizes this correctly and knows which icon layer to use.

I am hoping to put together a package in the next day or two which contains the scripts, SVGs, and all of the converted ICO icons for Dark Mode and Light Mode, in both regular FluentUI and filled icon sets.

**Comment by donho** [comment]
@dinkumoil 
Thank you for your info. What's the structure of your `ToolbarIcon`?

**Comment by dinkumoil** [comment]
@donho 

Not sure what you mean. Can you please explain in more detail, please?

**Comment by donho** [comment]
@dinkumoil 

I'm talking about this:
```
		struct toolbarIcons {
			HBITMAP	hToolbarBmp;
			HICON	hToolbarIcon;
		};
```
Ref: https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/MISC/PluginsManager/Notepad_plus_msgs.h#L148

When you use `NPPM_ADDTOOLBARICON` to send your ToolbarIcon object, your ToolbarIcon object should contain 1 bitmap & 1 icon. I just want to make sure that you pass both formats.

**Comment by dinkumoil** [comment]
@donho 

Thank you for your explanation. Currently I pass in `hToolbarBmp` a `HBITMAP` of a 256 colors (8 bits per pixel) palletized `BMP` image of 16x16 pixels in size and in `hToolbarIcon` the value `0` and it works. When I tried it vice versa (`0` as `hToolbarBmp` and a `HICON` as `hToolbarIcon`) it did not work.

Are there any requirements to the format of the icon whose handle is passed in `hToolbarIcon`? For example:

* Color depth (number of bits per pixel)
* Size
* Other requirements

It would be interesting to know not only the current requirements but also the future ones after all this dark mode stuff is integrated into Npp.

**Comment by donho** [comment]
@dinkumoil 
The `Record` in Pascal of your ToolbarIcon object should contain not only bitmap, but also icon.
However, I just checked the code source of Notepad++ and found out that only the bitmap format is processed; the icon format is ignored. I'll do a fix for that. In the future, plugins should send a structure (Record) of 4 object for ToolbarIcon - 1 bitmat (as today) and 2 additional icons (normal & disabled).

**Comment by dinkumoil** [comment]
@donho 

> In the future, plugins should send a structure (Record) of 4 object for ToolbarIcon - 1 bitmat (as today) and 2 additional icons (normal & disabled).

This presumably will cause crashes of lots of existing plugins when loaded by a Npp version including these changes. This is because new Npp versions will expect a longer struct than plugins will send and will read invalid icon handles from the additional struct members.

Since currently only bitmaps work as icon images, it can be assumed that plugin authors only use member `hToolbarBmp` of `struct toolbarIcons` and set member `hToolbarIcon` to `0`. Thus I would like to suggest that Npp handler code for `NPPM_ADDTOOLBARICON` checks, if member `hToolbarIcon` of the struct is set to `0`. If this is the case, the plugin has sent an old version of the struct and accessing the new members should be avoided.

**Comment by WildByDesign** [comment]
Dark Mode NppFluentUI icons are complete and packages to download within this comment. I have completely both the DarkMode icons and the DarkModeFilled icons and both sets have icons that contain both 32x32 and 16x16 pixel icons for maximum sharpness at both sizes without having to do any scaling.

Light Mode NppFluentUI icons are complete, however, I have not packaged them yet with this bundle today. All of the Light Mode SVGs are complete and look fantastic, but I still have to combine the 32x32 and 16x16 pixel icons into single icon files. I will try my best to package the Light Mode icons tomorrow.

These icons are sharp now and I am finally satisfied.

**Download:** 
[NppFluentUI-2021-04-30.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6408839/NppFluentUI-2021-04-30.zip)

**Screenshots:** (click to view at full size)

![NppFluentIcons-2021-04-30a](https://user-images.githubusercontent.com/26308319/116770310-da15ea00-aa10-11eb-88fb-0456251f8e4a.png)
![NppFluentIcons-2021-04-30b](https://user-images.githubusercontent.com/26308319/116770311-daae8080-aa10-11eb-95a8-93035a76bd07.png)

DarkMode and DarkModeFilled are ready to be added to Notepad++. I personally do not have to programming skills to do that part. LightMode and LightModeFilled icon sets should be ready tomorrow.

**Comment by WildByDesign** [comment]
Light Mode icons are complete now and ready to add to Notepad++. All toolbar button icon sets are complete.

**Download:** 
[NppFluentUI-2021-05-01.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6409860/NppFluentUI-2021-05-01.zip)
_*download package contains all completed DarkMode and LightMode icon sets._

**Screenshots:** (click to view at full resolution)

![NppFluentIcons-2021-05-01a](https://user-images.githubusercontent.com/26308319/116787652-4b868480-aa73-11eb-80dc-9136fafb701f.png)
![NppFluentIcons-2021-05-01b](https://user-images.githubusercontent.com/26308319/116787653-4c1f1b00-aa73-11eb-847c-baa7bd44e7e9.png)

I finally fixed the icon shading for Light Mode specifically for icons like Save, Save All, Redo, Undo, etc.

**Next Steps:**

- I need to figure out the best way for me to share the SVGs, scripts, etc. Is making a Github repo for them the best idea?
Days/weeks/hours went into the SVGs, so those things are like gold for future potential usage.

- I will continue to support these icon sets in Npp and therefore I am happy to update icons, fix inconsistencies, etc.

- My next steps will also involve converting other FluentUI icons into other areas of Npp UI for use in both Dark Mode and Light Mode. For example, I need to redo the icons for the tabs since there have been some changes to the dark mode theming that require changes on my end. I also need to create icons for the Light Mode tabs.

- Then I will look into other dialogs in Npp UI that have icons which need converting.

**Comment by donho** [comment]
@dinkumoil 
Plugin should provided both icon & bmp format for 1 added button. Currently it works only with bmp - I'm trying to make it work in the future. I'll let you know when I unblock this situation.

@WildByDesign 
Thank you for the beautiful, crystal clear icons. I'm still working on the PR - I'll merge them once I solved the plugin icon adding
on toolbar in Dark mode issue.

@mere-human 
I did some hack but don't figure out how to add an icon on Image list. Have you (or anyone) any idea what should I add in code to make it work? I'm working on this branch of Notepad++: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9808. Where the ico to bmp or vice verssa: https://github.com/donho/notepad-plus-plus/blob/new_monochromatic_icons/PowerEditor/src/WinControls/ToolBar/ToolBar.cpp#L286
Adding icon test with PluginDemo :  https://github.com/npp-plugins/plugindemo

**Comment by dinkumoil** [comment]
@donho 

More important: What do you think about my concerns regarding future extended format of struct `ToolbarIcon` when it consists of 4 objects instead of 2? Maybe you should [>> re-read my comment above <<](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9742#issuecomment-829377392).

**Comment by donho** [comment]
@dinkumoil 
Yes, I see your point/idea. The thing is, I cannot make toolbar icon work with plugin message `NPPM_ADDTOOLBARICON` at all so far - without making it work all plugin toolbar icons cannot be displayed in dark mode.
We can always consider to have a new message for preventing from crash once we make ico format work.

**Comment by mere-human** [comment]
> I did some hack but don't figure out how to add an icon on Image list. Have you (or anyone) any idea what should I add in code to make it work?

I'm not very familiar with image lists.
But, maybe it should be something like this?
```
HIMAGELIST hImgLst = (HIMAGELIST)::SendMessage(_hSelf, TB_GETIMAGELIST, 0, 0);
for (size_t j = 0; j < _nbDynButtons; ++j)
{
  ImageList_AddIcon(hImgLst, _vDynBtnReg.at(j)._hIcon);
  ```
I wasn't able to make it work. The button is just blank without an icon.

**Comment by donho** [comment]
@mere-human 
> I wasn't able to make it work. The button is just blank without an icon.

Thank you for trying...  Can you see the button? Could you send me the screenshot please?

**Comment by mere-human** [comment]
> Could you send me the screenshot please?

![image](https://user-images.githubusercontent.com/9664141/116898143-b0daa100-ac3e-11eb-8e06-65f59fd5f5ec.png)

@donho No icon at all.

**Comment by donho** [comment]
@mere-human 
It's bitmap part and it does work (at least we can see a button added).
Could you switch to the icons :
![image](https://user-images.githubusercontent.com/90293/116899734-373fb480-ac38-11eb-9d7e-551ab7279658.png)
to see if there's an added button please?

**Comment by namazso** [comment]
I haven't been following the dark mode issues recently, and I know this is kind of an "extra" request, but would it be possible to make sure that the whole dark mode thing and the icons work well on custom, dark themed windows? (ie everything will keep using or can be made to use system colors)
This is how it looks on pre-dark theme notepad++ (aka "light" theme) with a dark custom themed windows:
![image](https://user-images.githubusercontent.com/8676443/116900409-fe540f80-ac38-11eb-9209-89fbaeba2117.png)
Also, most issues affecting will also affect High Contrast users which might be an actual accessibility matter.

**Comment by mere-human** [comment]
> Could you switch to the icons :

No icon and no button:
![image](https://user-images.githubusercontent.com/9664141/116901432-960a2b80-ac42-11eb-8d63-ce39ce1a7c78.png)

```
Notepad++ v7.9.5   (64-bit)
Build time : May  3 2021 - 10:51:52
Path : E:\...\Notepad++.exe
Command Line : -nosession -multiInst
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 2009
OS Build : 19042.928
Current ANSI codepage : 1252
Plugins : NppPluginDemo.dll 

```

**Comment by donho** [comment]
Thank you @mere-human

**Comment by donho** [comment]
@namazso 
> I haven't been following the dark mode issues recently, and I know this is kind of an "extra" request, but would it be possible to make sure that the whole dark mode thing and the icons work well on custom, dark themed windows? (ie everything will keep using or can be made to use system colors)

Yes, that's what I'm working on.

**Comment by donho** [comment]
@WildByDesign 
For the Tabbar dark mode, the "readonly.ico" & "monitoring.ico" are lacked.
![image](https://user-images.githubusercontent.com/90293/117348175-32823700-aeaa-11eb-9887-eb1a1be1dfaa.png)

![image](https://user-images.githubusercontent.com/90293/117348369-68272000-aeaa-11eb-8552-73f55361c66a.png)

Could you provide these 2 icons please?

**Comment by mere-human** [comment]
![image](https://user-images.githubusercontent.com/9664141/117438984-f0550600-af3a-11eb-8f0e-2c64942a9310.png)
I tried the new icon set in a recent build with dark mode enabled.
The icons look very dim. Is it going to be fixed? Perhaps, we should invert the icons somehow, when in the dark mode.

**Comment by WildByDesign** [comment]
@donho Yes, absolutely. I created missing tab bar icons late last night and will review them today first and will attach them here as long as they turned out good.

@mere-human It appears that the latest Npp commits are using the light mode icons while in dark mode so what you are seeing is not intended. I provided light mode and dark mode icons, but only the light mode icons seem to have been added so far. 

Adding the dark mode icons would double the amount of icons. So I believe that Don and @adzm are trying to figure out how to programmatically invert the icons because that would be most ideal and more efficient.

**Comment by ArkadiuszMichalski** [comment]
@donho
One thing:
1. Open Notepad++ (default icons are `Standard icons: small`).
2. Go to `Dark Mode` tab and check `Enable dark mode`.
3. Our icons was changed to `Fluent UI: small`.
4. Back to `General` tab and see that we have still selected `Standard icons: small`.

Restarting the program updates these controls (or selecting them manually).

![image](https://user-images.githubusercontent.com/2730894/117673496-20a2db80-b1ab-11eb-8c3a-17fa38be48f2.png)

**Comment by namazso** [comment]
@donho 

Is https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9742#issuecomment-831369541 still being worked on? Latest snapshot build shows a pretty serious regression compared to the previous state:

![image](https://user-images.githubusercontent.com/8676443/117676660-0ae2e580-b1ae-11eb-9ff3-c8f19298a38c.png)

Also affecting high contrast users:

![image](https://user-images.githubusercontent.com/8676443/117677248-952b4980-b1ae-11eb-9bad-e4f1f677f640.png)

**Comment by donho** [comment]
@ArkadiuszMichalski 
Fixed in https://github.com/notepad-plus-plus/notepad-plus-plus/commit/c230ccf1bd74831a9483787a9c30c1792583074d

**Comment by mere-human** [comment]
> > Latest snapshot build shows a pretty serious regression compared to the previous state:
> 
> Are you sure you have the latest commit in master? Has anyone-else the same issue?

For me it looks fine.
Here is how it look on my machine:
![image](https://user-images.githubusercontent.com/9664141/117781091-ee47bb80-b248-11eb-9cbe-01105376fdfb.png)

And this is with High Contrast ON:
![image](https://user-images.githubusercontent.com/9664141/117781235-16371f00-b249-11eb-94f7-352e6b3e450e.png)

This is a build from commit a0f2586102a2807e688ea51d149113ebff9a99cf .

**Comment by WildByDesign** [comment]
@donho I came across two different issues regarding tab bar icons that hinder future quality of tab bar icons. I will describe the issues below. Please let me know if you want me to create a separate issue(s) for these bugs.

1.  The tab bar status icons (saved, unsaved, read-only, monitoring, etc.) have a weird scaling issue in Dark Mode only. It's possible that @adzm may understand this better because I believe that it may have something to do with the dark mode tab bar code. 

In normal (light) mode, those icons display at 12x12 pixels in default mode and 18x18 pixels when the reduced button is unchecked to display the larger tab bar size. However, in dark mode, those status icons are display at 13x13 pixels and 18x18 pixels. 

This is causing the dark mode tab bar status icons to render kind of fuzzy regardless of what I do.

**Problem:** something in the dark mode tab bar code is causing the status icons to stretch by 1 pixel to 13x13. What part of the dark mode code would cause the status icons to stretch by 1 pixel?

2.  The tab bar icons for closing the tab (closeTabButton.bmp, etc.) are in BMP format. The standard icon (toolbar) set are also BMP, however, they automatically have that feature in which the bottom left pixel is used to create a transparent background for those toolbar icons. These tab bar icons do not seem to inherit that same BMP transparency trick that the toolbar BMP icons have.

Having said that, it would be much better if those tab bar close icons could load ICO icons instead of BMP. I would be able to create much better looking close button / hover icons that can be used across light theme, dark theme, etc. if I can do proper transparency. Is it possible to load ICO icons for the tab bar close buttons?

For example, if we could keep the same naming scheme:
```
closeTabButton.bmp
closeTabButton_hover.bmp
closeTabButton_inact.bmp
closeTabButton_push.bmp
```

would become:
```
closeTabButton.ico
closeTabButton_hover.ico
closeTabButton_inact.ico
closeTabButton_push.ico
```

In my own testing, I attempted to change:
```
IDR_CLOSETAB                    BITMAP    "icons/closeTabButton.bmp"
IDR_CLOSETAB_INACT              BITMAP    "icons/closeTabButton_inact.bmp"
IDR_CLOSETAB_HOVER              BITMAP    "icons/closeTabButton_hover.bmp"
IDR_CLOSETAB_PUSH               BITMAP    "icons/closeTabButton_push.bmp"
```

to 
```
IDR_CLOSETAB                    ICON    "icons/closeTabButton.ico"
IDR_CLOSETAB_INACT              ICON    "icons/closeTabButton_inact.ico"
IDR_CLOSETAB_HOVER              ICON    "icons/closeTabButton_hover.ico"
IDR_CLOSETAB_PUSH               ICON    "icons/closeTabButton_push.ico"
```

But my attempts failed and only loaded a blank icon space. Keep in mind, I am not a programmer and therefore I was just messing around trying for an easy fix.

@donho  What do you think about this proposed change in point number 2?

Also, regarding point number 1, do you know what could be causing the 1 pixel stretch?

Cheers!

**Comment by donho** [comment]
@WildByDesign 
I will do an investigation about mentioned point as soon as I can.

**Comment by donho** [comment]
@dinkumoil 
The issue has been fixed:
https://github.com/notepad-plus-plus/notepad-plus-plus/commit/9a2dcaa5f8eb678e2b0c6494c16a8b9df37f06bb

**Comment by donho** [comment]
@WildByDesign 
I have checked the way to load the dark mode icons and the other icons in light mode, they are loaded in the same way, so should have the same size (13 or 20): 
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Notepad_plus.cpp#L227

OTOH, The different pixel that you observed could due to the light mode icons are smaller, so that creates the 1 pixel of less visually.

> Having said that, it would be much better if those tab bar close icons could load ICO icons instead of BMP.

I don't consider this change currently, since the effect you described is not even barely visible (I made a screenshot and X2 and I see nothing different). OTOH, it's the easiest way programmatically to have ico than bmp (that's why the toolbar uses rather ico than bmp). Hope you understand.

**Comment by adzm** [comment]
Sorry, I've been preoccupied and have not been able to work on this at all
recently. I'll let you know when things settle down and I have some free
time to invest again, probably next week or so, but please don't wait on me
for anything at the moment.
